# 承诺 Promise

Promise 代表一个**异步**操作的最终完成(或失败)及其结果值。 

Promise 是 `ES6` 的特性，如果需要兼容没有原生 Promise 的浏览器，只需引入一个 [Promise 库](https://github.com/stefanpenner/es6-promise) 就好了。

Promise 有三种可能的状态 

| 状态       | 说明                |
|-----------|---------------------|
| `pending` | 未定 `初始值`         | 
|`resolved` | 兑现 `resolve(value)`|
| `rejected`| 拒绝 `reject(error)` |

## 许下承诺 new Promise(executor)
许下一个承诺
```javascript
> let promise = new Promise(function(resolve, reject) {})
→ undefined

> promise
→ Promise {[[PromiseStatus]]: "pending", [[PromiseValue]]: undefined}
```

许下一个承诺并马上兑现 `resolve('战狼2')`
```javascript
> let 一起看电影 = new Promise(function(resolve, reject) { resolve('战狼2'); })
→ undefined

> 一起看电影
→ Promise {[[PromiseStatus]]: "resolved", [[PromiseValue]]: "战狼2"}
```

许下一个承诺，想了一秒钟，拒绝。`reject('今晚要上课')`
```javascript
> let 一起打篮球 = new Promise(function(resolve, reject) { setTimeout(() => reject('今晚要上课'), 1000); })
→ undefined

× Uncaught (in promise) 今晚要上课

> 一起打篮球
→ Promise {[[PromiseStatus]]: "rejected", [[PromiseValue]]: "今晚要上课"}
```
💡 `resolve` 和 `reject` 是 Promise 提供的函数。

💡 `new Promise(executor)` 中的 `executor` 函数在 Promise 初始化的时候立即执行。

## 然后怎样 then()
Promise 的值无法直接获取，需要调用 `then()` 方法。
```javascript
let trip = new Promise((resolve, reject) => {
  setTimeout(() => resolve('香港 🇭🇰'), 1000);
});

let hk = trip.then(function(place) {
  return place;
});

let au = hk.then(place => {
    console.log(place);                 // 香港 🇭🇰
    return `${place} -> 泰国 🇹🇭`;        
  })
  .then(places => {
    console.log(places);                // 香港 🇭🇰 -> 泰国 🇹🇭
    return `${places} -> 新加坡 🇸🇬`;
  })
  .then(places => {
    console.log(places);                // 香港 🇭🇰 -> 泰国 🇹🇭 -> 新加坡 🇸🇬
    return new Promise((resolve, reject) => {
      setTimeout(() => resolve(`${places} -> 新西兰 🇳🇿`), 3000);
    });
  })
  .then(places => {
    console.log(places);                // 香港 🇭🇰 -> 泰国 🇹🇭 -> 新加坡 🇸🇬 -> 新西兰 🇳🇿
    return new Promise((resolve, reject) => {
      setTimeout(() => resolve(`${places} -> 澳洲 🇦🇺`), 1500);
    });
  })
  .then(places => {
    console.log(places);                // 香港 🇭🇰 -> 泰国 🇹🇭 -> 新加坡 🇸🇬 -> 新西兰 🇳🇿 -> 澳洲 🇦🇺
    return places;
  });
```

## 如果不行 catch()
```javascript
let play = new Promise(function(resolve, reject) { 
  setTimeout(() => reject(new Error('不好意思😅，今晚要上课')), 1000);     // 通常用 Error 表示拒绝原因
});

play.catch(function(error) {
  console.log(error.message);
});
```
事实上，`catch()` 是 `then(null, onReject)` 的缩写。
```javascript
let home = au.then(() => {
  return Promise.reject(new Error('钱💰花🌺光了，肥家！'));
}).then(null, error => {
  console.log(error.message);
  return '家🏡';
});
```

## Fetch
`fetch()` 即 AJAX + Promise，详见 `JavaScript/AJAX/03. Fetch.md`。

### XMLHttpRequest + Promise
```javascript
function getJSON(url) {
  return new Promise((resolve, reject) => {
    let xhr = new XMLHttpRequest();
    xhr.responseType = 'json';
    xhr.open('GET', url);
    xhr.onload = function() { resolve(this.response) };
    xhr.onerror = reject;
    xhr.send();
  });
}

getJSON('https://fe13.now.sh/api/posts').then(console.log)
```

## 隐式 try...catch
`new Promise(executor)` 的 `executor` 被一个无形的 `try...catch` 包围，一旦出错，Promise 的状态变成 `rejected`。

```javascript
new Promise(() => {
  throw new Error('要迟到啦啦啦啦~\(≧▽≦)/~⏰⏰⏰⏰⏰');
}).catch(error => {
  console.log(error.message);
  return '今天星期天';
}).then(day => {
  console.log(day, '继续睡觉😴');
})
```

## Promise API
`Promise.resolve()` 方法返回一个已经兑现的承诺😁。
```javascript
> Promise.resolve('去现场看 NBA 总决赛');
→ Promise {[[PromiseStatus]]: "resolved", [[PromiseValue]]: "去现场看 NBA 总决赛"}
```

`Promise.reject()` 方法返回一个已经被拒绝兑现的承诺😅。
```javascript
Promise.reject(new Error('怎么了 你累了 说好的幸福呢'))
  .catch(error => {
    console.log(error.message);
    return '我懂了 不说了 爱淡了 梦远了'
  })
  .then(byebye => {
    console.log(byebye);
  })
```
 
`Promise.all()` 方法返回一个 Promise，该 Promise 只在一系列指定的 Promise 被 `resolve` 后才会被 `resolve`。
```javascript
Promise.all([
  fetch('https://raw.githubusercontent.com/fe13/fe/master/JavaScript/AJAX/json/jay.json'),
  fetch('https://raw.githubusercontent.com/fe13/fe/master/JavaScript/AJAX/json/jay.albums.json')
])
  .then(responses => Promise.all(responses.map(response => response.json())))
  .then(jsons => {
    let [jay, albums] = jsons;     // 等价于 let jay = jsons[0], albums = jsons[1];
    jay.albums = albums;
    console.log(jay);
    return jay;
  })
  .catch(error => {
    console.log(error.message);    // 其中一个请求出现错误整个 Promise 就会被 reject
  })
```
💡 `response.json()` 返回一个 Promise。

`Promise.race()` 方法与 `Promise.all()` 类似，但它关注的是第一个被 `resolve` 或 `reject` 的 Promise。

```javascript
Promise.race([
  new Promise((resolve, reject) => setTimeout(() => resolve('🚒'), 9000)),
  new Promise((resolve, reject) => setTimeout(() => resolve('🚀'), 1000)),
  new Promise((resolve, reject) => setTimeout(() => reject(new Error('💥')), 3000))
]).then(console.log)    // 🚀
```

## async/await
`async` `await` 让 Promise 用起来更舒服，而且很容易理解和使用。

### async
```javascript
async function animal() {
  return '🐶';
}
```
`async` 表明该函数返回一个 Promise，若返回的是非 Promise 的值，则自动转成 `return Promise.resolve(val)`。

```javascript
> animal()
→ Promise {[[PromiseStatus]]: "resolved", [[PromiseValue]]: "🐶"}
```
### await
`await` 等待一个 Promise 被 `resolve` 并返回它的值。
```javascript
(async function() {
  let url = 'https://raw.githubusercontent.com/fe13/fe/master/JavaScript/AJAX/json/jay.json';
  let res = await fetch(url);
  let jay = await res.json();
  console.log(jay);
})();
```
`async` `await` 也可以在方法上使用。
```javascript
class Cat {
  async sleep(sec) {
    console.log('😴 💤💤💤');
    await new Promise(resolve => setTimeout(() => resolve(), sec * 1000));
    console.log('😎 🌈🌈🌈');
  }
}

let tom = new Cat();
tom.sleep(5);
```
`await` 还可以结合 `Promise.all()` 使用。
```javascript
(async function() {
  let responses = await Promise.all([
    fetch('https://raw.githubusercontent.com/fe13/fe/master/JavaScript/AJAX/json/jay.json'),
    fetch('https://raw.githubusercontent.com/fe13/fe/master/JavaScript/AJAX/json/jay.albums.json')
  ]);
  let [jay, albums] = await Promise.all(responses.map(response => response.json()));
  jay.albums = albums
  console.log(jay);
})();
```

### 错误处理
使用 `async` `await` 很简单，但还是需要处理错误。
```javascript
async function getJSON(url) {
  try {
    let res = await fetch(url);
    return await res.json();
  } catch (err) {
    console.log(err);
  }
}
```

## 知道更好
`$.ajax()` 返回一个 `thenable` 对象(就是有 `then()` 方法的对象)，可以被当成一个 Promise。

`then(onresolve, onreject)` `catch(onreject)` 的中 `onresolve` 和 `onreject` 总是异步执行的。

```javascript
let promise = new Promise((resolve, reject) => resolve('🚕'));

promise.then(console.log);

console.log('🚀');       // 🚀 先被打印出来
```

`unhandledrejection` 事件
```javascript
window.addEventListener('unhandledrejection', function (event) {
  console.warn(event.reason);
});
```

## 视频教程
* https://www.youtube.com/watch?v=vQ3MoXnKfuQ

## 参考案例
* https://twhy.github.io/github-user-finder [源码](https://github.com/twhy/github-user-finder)

## 开源项目
* [axios](https://github.com/mzabriskie/axios) Promise based HTTP client for the browser and node.js

## 参考链接
* [Promise A+ 规范](https://promisesaplus.com)
* http://callbackhell.com
* https://davidwalsh.name/promises
* https://github.com/mzabriskie/axios
* http://javascript.info/promise-basics
* http://javascript.info/promise-api
* http://javascript.info/promise-chaining
* http://javascript.info/async-await
* https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API
* https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Using_promises
* https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise
* https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all
* https://developers.google.com/web/fundamentals/getting-started/primers/promises
* https://developer.mozilla.org/en-US/docs/Web/Events/unhandledrejection
* https://stackoverflow.com/questions/32721850/why-the-response-object-from-javascript-fetch-api-is-a-promise
